### PR TITLE
[tde] Add readme and improvements for contrib/dynamic_embedding

### PR DESCRIPTION
**Pull request by zhuzilin**
This PR adds `README.md` and multiple bugfixs and performance enhancements for the dynamic embedding extension in the `contrib/dynamic_embedding` folder. The relevant bugs or performance enhancement are already fixed in the top level code (`torchrec/csrc`) during migration.

Thank you for your time on reviewing this PR.

gently ping @colin2328

### CODE REVIEW

File: contrib/dynamic_embedding/src/torchrec_dynamic_embedding/distributed/comm.py [hunk 1]
```diff
@@ -1,127 +1,91 @@
-from dataclasses import dataclass
 from typing import List, Optional
```
Comment (colin2328) [inline contrib/dynamic_embedding/src/torchrec_dynamic_embedding/distributed/comm.py:1]: let's put this into torchrec/distributed/dynamic_embedding/comm.py?

File: contrib/dynamic_embedding/README.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
+# TorchRec Dynamic Embedding
```
Comment (colin2328) [inline contrib/dynamic_embedding/README.md:1]: let's move this into torchrec/distributed/dynamic_embedding?

File: contrib/dynamic_embedding/README.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        self.emb1 = EmbeddingCollection(tables=config1, device=torch.device("meta"))
+        self.emb2 = EmbeddingCollection(tables=config2, device=torch.device("meta"))
+        ...
+
+    def forward(self, kjt1, kjt2):
+        ...
+
+m = Model(config1, config2)
+m = DistributedModelParallel(m)
+dataloader = tde.wrap(
```
Comment (colin2328) [inline contrib/dynamic_embedding/README.md:38]: let's make this into a train_pipeline? (see torchrec/distributed/train_pipelin.py)

**Review by colin2328 — COMMENTED** [review]
High level thoughts:
let's move the user facing code (i.e python code) outside of contrib/dynamic_embedding and into torchrec/distributed/dynamic_embedding?

THen, we can put user facing python modules w/ docstrings that call the _modules that are binded w/ pybind (see See how torchtext does this here https://github.com/pytorch/text/blob/f653dac9cec31f061ad9fba42b9e2a20aca7f569/torchtext/prototype/vocab_factory.py#L4-L48 for an example)

let's move from tde.wrap() API to create a train_pipeline.py that can inherit from TrainPipelineBase?

Let's create a dynamicEmbeddingSharder to shard EmbeddingCollections and create DynamicEmbeddingCollections

WDYT?

**Comment by zhuzilin** [comment]
@colin2328 
> let's move from tde.wrap() API to create a train_pipeline.py that can inherit from TrainPipelineBase?
>
> Let's create a dynamicEmbeddingSharder to shard EmbeddingCollections and create DynamicEmbeddingCollections

I agree. A `DynamicEmbeddingCollections` will make the user experience much more coherent. And we can learn from the new `TrainPipelineSparseDist` to pipeline the data processing and the model training. The only issue is that the currrent `TrainPipeline`s are using GPU streams for pipeline, while we may need multithreading for dynamic embedding, as the data processing are currently on CPU.

**Comment by colin2328** [comment]
stamping PR, thanks @zhuzilin .
Please feel free to move around, I'll work on DynamicEmbeddingCOllection API next week

**Pull request closed without merging**